### ISSUE TITLE
Limit Role Access for specific Tables

### ISSUE THREAD

**Issue by pjuanda** [apache/incubator-superset#8879]
**Scenario:**
Let's say i have database called `northwind`. Inside db, i have 2 tables, `customer`  and `employee`. I want to create role for `Human Resource` and  `Marketing` . Users with role `Human Resource` only can run query via SQL Lab to table `employee` and can't query from table `customer` and vice versa. Users will have role `Gamma` and `sql_lab` as base roles.

**What i've done so far is:**
to run query on SQL Lab, one must have access to database hence i have to grant `database access on northwind`, but if i grant this permission, the user can run query on both tables `employee` and `customer`

**Comment by hemantaggarwal** [comment]
In SQL Lab its not possible to restrict access on table level. I was also looking for this but I think currently this functionality is not there. We can restrict access to schema level.

**Comment by ArcherTsai** [comment]
Someone has raised issues about this before and it seems no one tried to deal with this.
https://github.com/apache/incubator-superset/issues/1531

I am not sure this is a bug or a feature. There are some codes related to access control in SQL Lab. But it would bypass the table access permission check and grant access permission to all tables in the DB as long as a user have access permission to the DB. 

I have done some modifications to fix this. And maybe I can commit it a few days later.

**Comment by mistercrunch** [comment]
Oh right. This is a regression since at some point people didn't want to list the databases that users don't have access to. By restricting this now you cannot see the database in the list.

The logic defining that list is here:
https://github.com/apache/incubator-superset/blob/003e98c231801b0ea34ebaefcff06c7ae9568faf/superset/views/database/mixins.py#L43-L51

One way to fix this is to figure out which tables the user has access to, and add a close in the or_ that essentially says `OR id IN (SELECT database_id FROM sqla_table WHERE perm IN (...{more subqueries in the security related tables}...))`

**Comment by f7753** [comment]
Any progress? Looking forward to this fix, or what can I do to fix this?

**Comment by f7753** [comment]
@ArcherTsai  

> Someone has raised [a issue](https://github.com/apache/incubator-superset/issues/1531) about this before and it seems no one tried to deal with this.
> 
> I am not sure this is a bug or a feature. There are some codes related to access control in SQL Lab. But it would bypass the table access permission check and grant access permission to all tables in the DB as long as a user have access permission to the DB.
> 
> According to the descriptions of [SQL Lab](https://superset.incubator.apache.org/security.html#sql-lab), access is given on a database basis. So maybe it is the design of Superset.
> 
> I have done some modifications to fix this. And maybe I can commit it a few days later.

Hi，Archer, thanks for your contributition, any progress been made? Looking forward to this feature(or fix) , is there anything I can do to accelerate this issue?

**Comment by ArcherTsai** [comment]
@f7753 

Sorry for the late reply. 

What I modified is in superset/security.py in branch 0.35
https://github.com/apache/incubator-superset/blob/7cb3d3b5d2e05f98f0683bc72514e4d9f4ce13f8/superset/security.py#L291-L318

As I noticed that this is different from the code pointed by mistercrunch and I don't have time to dig into the code, I did not commit my change.

Also, I noticed that superset/security.py does not exist in 0.36 branch. Maybe this is the best modification for this issue.
